lib/functional.fnl

```diff
@@ -163,14 +172,34 @@
    tbl
    paths))
 
-(fn map
-  [f tbl]
-  (reduce
-    (fn [new-tbl v k]
-      (table.insert new-tbl (f v k))
-      new-tbl)
-    []
-    tbl))
+(fn zip [...]
+  "Groups corresponding elements from multiple lists into a new list, truncating at the length of the smallest list."
+  (let [tbls [...]
+        result []]
+    (if (= 1 (length tbls))
+        (table.insert result (. tbls 1))
+        (let []
+          (for [idx 1 (length (. tbls 1))]
+            (let [inner []]
+              (each [_ tbl (ipairs tbls) &until (not (. tbl idx))]
+                (table.insert inner (. tbl idx)))
+              (table.insert result inner)))))
+    result))
+
+(fn map [f ...]
+  (let [args [...]
+        tbls (zip (table.unpack args))]
+    (if (= 1 (count args))
+        (icollect [_ v (pairs (first args))]
+          (apply f v))
+        (accumulate [acc []
```

Comment:
We can maybe remove our `reduce` and use this instead?

Reply:
There are some places where reduce as a function has its value but I think replacing it with a thin wrapper around `accumulate` should be fine

Reply:
agreed. Working on it

Reply:
#194